profiler/profile.go

```diff
@@ -44,48 +44,125 @@ const (
+// map below.
+type collector struct {
+	// Name is used to lookup the profile for pprof.Lookup(name) and returned by
+	// ProfileType.String(). For profile types that don't use this approach, the
+	// name can be chosen arbitrarily.
+	Name string
+	// Filename is the filename used for uploading the profile to the datadog
+	// backend which is aware of them. Delta profiles are prefixed with "delta-"
+	// automatically.
+	Filename string
+	// Delta enables delta profiling for pprof profiles. This is useful for
+	// profiles that are represent samples collected over the lifetime of the
+	// process such as heap, block, mutex, etc..
+	Delta bool
+	// Collect collects the given profile and returns the data for it. Most
+	// profiles will be in pprof format, i.e. gzip compressed proto buf data.
+	Collect func(collector, *profiler) ([]byte, error)
+}
+
+// collectors is a 1:1 map between ProfileType and collector implementations.
+var collectors = map[ProfileType]collector{
+	CPUProfile: {
+		Name:     "cpu",
+		Filename: "cpu.pprof",
+		Collect: func(_ collector, p *profiler) ([]byte, error) {
+			var buf bytes.Buffer
+			if err := startCPUProfile(&buf); err != nil {
+				return nil, err
+			}
+			time.Sleep(p.cfg.cpuDuration)
+			stopCPUProfile()
+			return buf.Bytes(), nil
+		},
+	},
+	HeapProfile: {
+		Name:     "heap",
+		Filename: "heap.pprof",
+		Delta:    true,
+		Collect:  collectGenericProfile,
+	},
+	MutexProfile: {
+		Name:     "mutex",
+		Filename: "mutex.pprof",
+		Delta:    true,
+		Collect:  collectGenericProfile,
+	},
+	BlockProfile: {
+		Name:     "block",
+		Filename: "block.pprof",
+		Delta:    true,
+		Collect:  collectGenericProfile,
+	},
+	// TODO(fg) enable Delta for this? could be cool to see newly created
+	// goroutines.
+	GoroutineProfile: {
+		Name:     "goroutine",
+		Filename: "goroutines.pprof",
+		Collect:  collectGenericProfile,
+	},
+	// TODO(fg) enable Delta for this? could be cool to goroutines that
+	// recently entered waiting states.
+	expGoroutineWaitProfile: {
+		Name:     "goroutinewait",
+		Filename: "goroutineswait.pprof",
+		Collect: func(c collector, _ *profiler) ([]byte, error) {
+			var (
+				now   = now()
+				text  = &bytes.Buffer{}
+				pprof = &bytes.Buffer{}
+			)
+			if err := lookupProfile(c.Name, text, 2); err != nil {
+				return nil, err
+			}
+			err := goroutineDebug2ToPprof(text, pprof, now)
+			return pprof.Bytes(), err
+		},
+	},
+	MetricsProfile: {
+		Name:     "metrics",
+		Filename: "metrics.json",
+		Collect: func(_ collector, p *profiler) ([]byte, error) {
+			var buf bytes.Buffer
+			err := p.met.report(now(), &buf)
+			return buf.Bytes(), err
+		},
+	},
+}
+
+func collectGenericProfile(c collector, _ *profiler) ([]byte, error) {
+	var buf bytes.Buffer
+	err := lookupProfile(c.Name, &buf, 0)
+	return buf.Bytes(), err
+}
+
+// collector returns the collector holding the implementation details of the
+// profile type.
```

Comment:
```suggestion
// collector returns the collector for this profileType.
```
A bit clearer and simpler imo.

Reply:
👍